include/ismrmrd/dataset_dtypes.h

```diff
@@ -189,6 +184,174 @@ template <> DataType get_hdf5_data_type<AcquisitionHeader_with_data<int32_t> >()
+  dtype.insertMember("image_series_index",    HOFFSET(ImageHeader, image_series_index),    PredType::NATIVE_UINT32);
+
+  dims[0] = ISMRMRD_USER_INTS;
+  DataType usints_array_type = ArrayType(PredType::NATIVE_INT32, 1, &dims[0]);
+  dtype.insertMember("user_int", HOFFSET(ImageHeader, user_int), usints_array_type);
+
+  dims[0] = ISMRMRD_USER_FLOATS;
+  DataType usfloats_array_type = ArrayType(PredType::NATIVE_FLOAT, 1, &dims[0]);
+  dtype.insertMember("user_float",           HOFFSET(ImageHeader, user_float),           usfloats_array_type);
+  dtype.insertMember("attribute_string_len", HOFFSET(ImageHeader, attribute_string_len), PredType::NATIVE_UINT32);
+
+  return dtype;
+}
+
+/* Image data */
+template <> DataType get_hdf5_data_type<std::complex<double> >()
+{
+  hid_t datatype = H5Tcreate(H5T_COMPOUND, sizeof(std::complex<double>));
+  herr_t h5status = H5Tinsert(datatype, "real", 0, H5T_NATIVE_DOUBLE);
+  if (h5status < 0)
+  {
+    throw std::runtime_error("Failed get complex float data type");
+  }
+  h5status = H5Tinsert(datatype, "imag", sizeof(double), H5T_NATIVE_DOUBLE);
+  if (h5status < 0)
+  {
+    throw std::runtime_error("Failed get complex float data type");
+  }
+  return datatype;
+}
+
+template <> DataType get_hdf5_data_type<std::complex<float> >()
+{
+  hid_t datatype = H5Tcreate(H5T_COMPOUND, sizeof(std::complex<float>));
+  herr_t h5status = H5Tinsert(datatype, "real", 0, H5T_NATIVE_FLOAT);
+  if (h5status < 0)
+  {
+    throw std::runtime_error("Failed get complex float data type");
+  }
+  h5status = H5Tinsert(datatype, "imag", sizeof(float), H5T_NATIVE_FLOAT);
+  if (h5status < 0)
+  {
+    throw std::runtime_error("Failed get complex float data type");
+  }
+  return datatype;
+}
+
+template <> DataType get_hdf5_data_type<uint16_t>()
+{
+  hid_t datatype = H5Tcopy(H5T_NATIVE_UINT16);
+  return datatype;
+}
+
+template <> DataType get_hdf5_data_type<int16_t>()
+{
+  hid_t datatype = H5Tcopy(H5T_NATIVE_INT16);
+  return datatype;
+}
+
+template <> DataType get_hdf5_data_type<uint32_t>()
+{
+  hid_t datatype = H5Tcopy(H5T_NATIVE_UINT32);
+  return datatype;
+}
+
+template <> DataType get_hdf5_data_type<int32_t>()
+{
+  hid_t datatype = H5Tcopy(H5T_NATIVE_INT32);
+  return datatype;
+}
+
+template <> DataType get_hdf5_data_type<float>()
+{
+  hid_t datatype = H5Tcopy(H5T_NATIVE_FLOAT);
+  return datatype;
+}
+
+template <> DataType get_hdf5_data_type<double>()
+{
+  hid_t datatype = H5Tcopy(H5T_NATIVE_DOUBLE);
+  return datatype;
+}
+
+/***********************************************************************************************************************
+ Data types for Waveform class
+ **********************************************************************************************************************/
+template <>
+DataType get_hdf5_data_type<WaveformHeader>()
+{
+  CompType dtype(sizeof(WaveformHeader));
+
+  dtype.insertMember("begin_time_stamp",  HOFFSET(WaveformHeader, begin_time_stamp),  PredType::NATIVE_UINT64);
```

Comment:
No need for `begin_time_stamp` and `end_time_stamp`. It should suffice to have `time_stamp` (in nano seconds) and `dwell_time_ns`. The `end_time_stamp` is equal to `time_stamp` + `dwell_time_ns` * `number_of_samples`